packages/lib/bookings/findQualifiedHosts.ts

```diff
@@ -1,38 +1,201 @@
+    .filter((host) => !newValue.includes(host))
+    .map((host) => ({
+      ...host,
+      disqualifyReason: reason,
+    }));
+  if (fallbackRRHosts.length > 0) {
+    return [...fixedHosts, ...fallbackRRHosts];
+  }
+  return [];
+};
+
+const isRoundRobinHost = <T extends { isFixed: boolean }>(host: T): host is T & { isFixed: false } => {
+  return host.isFixed === false;
+};
+
+export const findQualifiedHosts = async <
+  T extends {
+    email: string;
+    id: number;
+    credentials: CredentialPayload[];
+    userLevelSelectedCalendars: SelectedCalendar[];
+  } & Record<string, unknown>
+>({
+  eventType,
+  rescheduleUid,
+  routedTeamMemberIds,
+  contactOwnerEmail,
+}: {
   eventType: {
     id: number;
     maxLeadThreshold: number | null;
-    hosts?: ({ isFixed: boolean; createdAt: Date; priority?: number | null; weight?: number | null } & {
-      user: T;
-    })[];
+    hosts?: Host<T>[];
     users: T[];
     schedulingType: SchedulingType | null;
+    isRRWeightsEnabled: boolean;
     rescheduleWithSameRoundRobinHost: boolean;
-  },
-  isReschedule: boolean
-): Promise<
-  {
+  } & EventType;
+  rescheduleUid: string | null;
+  routedTeamMemberIds: number[];
+  contactOwnerEmail: string | null;
+}): Promise<{
+  qualifiedHosts: {
+    isFixed: boolean;
+    createdAt: Date | null;
+    priority?: number | null;
+    weight?: number | null;
+    user: T;
+  }[];
+  fallbackHosts?: {
     isFixed: boolean;
-    email: string;
     createdAt: Date | null;
     priority?: number | null;
     weight?: number | null;
     user: T;
-  }[]
-> => {
-  const { hosts, fallbackHosts } = getNormalizedHosts({ eventType });
-  const qualifiedHosts = hosts
-    ? await filterHostsByLeadThreshold({
-        eventTypeId: eventType.id,
-        hosts,
-        maxLeadThreshold:
-          isReschedule && eventType.rescheduleWithSameRoundRobinHost ? null : eventType.maxLeadThreshold,
-      })
-    : fallbackHosts;
-  return qualifiedHosts;
+  }[];
+}> => {
+  const { hosts: normalizedHosts, fallbackHosts: fallbackUsers } = getNormalizedHosts({ eventType });
+  // not a team event type, or some other reason - segment matching isn't necessary.
+  if (!normalizedHosts) {
+    return { qualifiedHosts: fallbackUsers };
+  }
+
+  const fixedHosts = normalizedHosts.filter((host) => host.isFixed);
+  const roundRobinHosts = normalizedHosts.filter(isRoundRobinHost);
+
+  const hostsAfterContactOwnerMatching = applyFilterWithFallback(
+    roundRobinHosts,
+    roundRobinHosts.filter((host) => host.user.email === contactOwnerEmail)
+  );
+
+  if (hostsAfterContactOwnerMatching.length === 1) {
+    // push all disqualified hosts into fallback, contact owner is recoverable
+    const fallbackHosts = getFallbackHosts(
+      fixedHosts,
+      roundRobinHosts,
+      hostsAfterContactOwnerMatching,
+      "NotContactOwner"
+    );
+
+    return {
+      qualifiedHosts: [...fixedHosts, ...hostsAfterContactOwnerMatching],
+      fallbackHosts,
```

Comment:
I think the fallbackHosts should never be the event type hosts. 

Fallback hosts should ideally be `hostsAfterFairnessMatching`. In case the contact owne has no availability for 2 weeks we would ignore all other logic (attribute routing, fairness,..)

Reply:
This fallbackHosts will only every be used when a CRM contact owner has been found, and this result is returned.

Reply:
@emrysal should be fixed with https://github.com/calcom/cal.com/pull/18550/commits/ae125188a9e4b7a0e594fdc00d1c777fd489950a